### PR TITLE
nixos/limesurvey: drop default encryption key and nonce

### PR DESCRIPTION
**Pull request by wegank**
## Description of changes

https://github.com/NixOS/nixpkgs/pull/222149#discussion_r1653731189

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by anthonyroussel** [comment]
@ofborg test limesurvey

File: nixos/modules/services/web-apps/limesurvey.nix [hunk 1]
```diff
@@ -183,6 +183,12 @@ in
       { assertion = cfg.database.createLocally -> cfg.database.passwordFile == null;
         message = "a password cannot be specified if services.limesurvey.database.createLocally is set to true";
       }
+      { assertion = cfg.encryptionKey != options.services.limesurvey.encryptionKey.default;
```
[the remark this answers is not in the document]
Comment (wegank) [reply nixos/modules/services/web-apps/limesurvey.nix:186]: Fixed.

**Comment by wegank** [comment]
@ofborg test limesurvey

**Review by anthonyroussel — APPROVED** [review]

**Comment by mweinelt** [comment]
Can we get rid of the default? What is the idea of doing an assertion instead? How does that affect backwards compatibility? Can deployments even swap out the encryption key?

**Review by LeSuisse — COMMENTED** [review]
Maybe it would interesting to drop the default values and take an `encryptionKeyFile` and `encryptionNonceFile` instead which would then be used to be loaded using systemd `LoadCredential` feature.

One thing I cannot figure out it's how existing instances can rotate their keys. It looks like it is not possible which is not great in our situation :/ 


Edit: Sorry for the double post, no realtime on the smartphone app :/

**Comment by mweinelt** [comment]
Can we use file_get_contents to make config.php look up the secrets from a path?

**Comment by wegank** [comment]
Since the whole configuration is passed via `json_decode('${builtins.toJSON cfg.config}', true)`, there's no direct way round, and I'm not at all sure that rewriting it with an `mkConfigFile` containing special handling for paths won't create more regression...

**Comment by LeSuisse** [comment]
We can manipulate the `limesurveyConfig` we create in order to load the config and the credentials without touching the generation. I have a working change here if you want to take a look: https://github.com/LeSuisse/nixpkgs/commit/7d14d4164fa88f10e873fcbbc999221e51892299

Not our main concern right now but the `services.limesurvey.database.passwordFile` setting is likely broken

**Comment by wegank** [comment]
Thanks, I'm now using your commit, with an extra trim for file_get_contents.

**Review by LeSuisse — APPROVED** [review]
> Thanks, I'm now using your commit, with an extra trim for file_get_contents.

Thanks! I did the same at some point while working on it but I forgot to include it in the change I published.

Anyway I believe the PR is in good shape now, I'm going to merge it. I do not think we will get more feedback.

I will take a look at `services.limesurvey.database.passwordFile` in a subsequent PR this week-end.

**Pull request merged**